modules/renter/downloadcache.go

```diff
@@ -14,21 +14,33 @@ import (
 // TODO this won't be necessary anymore once we have partial downloads.
 func (udc *unfinishedDownloadChunk) addChunkToCache(data []byte) {
 	if udc.download.staticDestinationType == destinationTypeSeekStream {
-		// We only cache streaming chunks since browsers and media players tend to only request a few kib at once when streaming data. That way we can prevent scheduling the same chunk for download over and over.
+		// We only cache streaming chunks since browsers and media players tend
+		// to only request a few kib at once when streaming data. That way we can
+		// prevent scheduling the same chunk for download over and over.
 		return
 	}
 	udc.cacheMu.Lock()
+	defer udc.cacheMu.Unlock()
+
 	// Prune cache if necessary.
-	// TODO insteado of deleting a 'random' key, delete the
-	// least-recently-accessed element of the cache.
-	for key := range udc.chunkCache {
-		if len(udc.chunkCache) < downloadCacheSize {
-			break
+	if len(udc.chunkCache) >= downloadCacheSize {
+		var oldestKey string
+		oldestTime := time.Now().Second()
+
+		for key := range udc.chunkCache {
+			if udc.chunkCache[key].timestamp.Second() < oldestTime {
+				oldestTime = udc.chunkCache[key].timestamp.Second()
+				oldestKey = key
+			}
 		}
-		delete(udc.chunkCache, key)
+		delete(udc.chunkCache, oldestKey)
+	}
+
+	cd := cacheData{
+		data:      data,
+		timestamp: time.Now(),
 	}
-	udc.chunkCache[udc.staticCacheID] = data
-	udc.cacheMu.Unlock()
+	udc.chunkCache[udc.staticCacheID] = cd
```

Comment:
you could save a line here if you didn't save the object in a variable like this:
```
udc.chunkCache[udc.staticCacheID] = cacheData {
    data: data,
    timestamp: time.Now(),
}
```